Kafka cluster Id in status

lga-zurich opened the pull request:
### Type of change

- Enhancement / new feature

### Description

As motivated in #4028, this PR introduces the Kafka cluster Id in the status of a Kafka CRD. 

The advantage of the Kafka cluster Id, as opposed to, say, URLs or bootstrap servers, is that all observers can agree on the same id for a given Kafka cluster, independently of how they access it, for example from within or outside the Kubernetes cluster. 

Having the cluster Id available in the status provides the information to any entity that can list the CR, without requiring higher privileges to access Zookeeper or the Kafka Admin API.

### Checklist

I'm unsure about some of these. The test doesn't do much beyond checking that the cluster Id being set in the status is actually what one gets out of it. Suggestions are welcomed. I have satisfactorily run the patched operator on Minikube. If there is documentation I would need to change I'll be glad to get a pointer.

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1874,6 +1879,32 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
+        Future<ReconciliationState> kafkaGetClusterId() {
+            String clusterId = null;
+            log.debug("{}: Attempt to get clusterId", reconciliation);
+            List<String> podNames = IntStream.range(0, kafkaCluster.getReplicas()).boxed().collect(Collectors.toList()).stream().map(podId -> KafkaCluster.kafkaPodName(this.name, podId)).collect(Collectors.toList());
+            try {
+                String bootstrapHostnames = podNames.stream().map(podName -> KafkaCluster.podDnsName(this.namespace, this.name, podName) + ":" + KafkaCluster.REPLICATION_PORT).collect(Collectors.joining(","));
```
You should be able to use the bootstrap service here. That should save you from constructing the name like this.

There is a class `KafkaResources` which has method `bootstrapServiceName` which should be able to give it to you. Or you can add to the `KafkaCluster` class some new method similar to `podDnsName` (since oyu need to use it with the namespace).

I do not think there is any advantage here in using the pod names => it just adds possible complications with DNS changes.

lga-zurich replied:
ok, I will check that. I did try to use `kafkaBootstrapDnsName` but in my tests it was always empty.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1874,6 +1879,32 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
+            try {
+                String bootstrapHostnames = podNames.stream().map(podName -> KafkaCluster.podDnsName(this.namespace, this.name, podName) + ":" + KafkaCluster.REPLICATION_PORT).collect(Collectors.joining(","));
+                log.debug("{}: Creating AdminClient for clusterId using {}", reconciliation, bootstrapHostnames);
+                Admin kafkaAdmin = adminClientProvider.createAdminClient(bootstrapHostnames, this.clusterCa.caCertSecret(), this.oldCoSecret, "cluster-operator");
```
Is `oldCoSecret` really the right one to use? If this is old, it sounds like sometimes it might not be valid anymore.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1874,6 +1879,32 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
+                log.debug("{}: Creating AdminClient for clusterId using {}", reconciliation, bootstrapHostnames);
+                Admin kafkaAdmin = adminClientProvider.createAdminClient(bootstrapHostnames, this.clusterCa.caCertSecret(), this.oldCoSecret, "cluster-operator");
+                clusterId = kafkaAdmin.describeCluster().clusterId().get();
+            } catch (KafkaException e) {
+                log.debug("{}: Kafka exception getting clusterId {}", reconciliation, e.getMessage());
+            } catch (Exception e) {
+                log.debug("{}: Exception getting clusterId {}", reconciliation, e.getMessage());
+            }
```
What are the situations when this happens? In general:
* Is it intentional to continue as success when it fails instead of failing the reconciliation?
* Should the error really be only on debug level?

Both of these IMHO suggest that this is expected to fail regularly while the failure does not indicate any problems. So I would like to understand why.

lga-zurich replied:
As I said above, I didn't want to make an error here critical. Either it is transient and it will eventually work or the set up is so screwed up that something else more fundamental will blow up. But that can be changed, of course.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1874,6 +1879,32 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
+                String bootstrapHostnames = podNames.stream().map(podName -> KafkaCluster.podDnsName(this.namespace, this.name, podName) + ":" + KafkaCluster.REPLICATION_PORT).collect(Collectors.joining(","));
+                log.debug("{}: Creating AdminClient for clusterId using {}", reconciliation, bootstrapHostnames);
+                Admin kafkaAdmin = adminClientProvider.createAdminClient(bootstrapHostnames, this.clusterCa.caCertSecret(), this.oldCoSecret, "cluster-operator");
+                clusterId = kafkaAdmin.describeCluster().clusterId().get();
```
I'm not too familiar with the Kafka AdminClient API. But the `get()` IMHO usually suggests it is a blocking call. So it should probably be wrapped into execute blocking or handled asynchronously?

lga-zurich replied:
Agreed, the KafkaFuture.get() will wait if necessary, that was just lazy.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1874,6 +1879,32 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
+                clusterId = kafkaAdmin.describeCluster().clusterId().get();
+            } catch (KafkaException e) {
+                log.debug("{}: Kafka exception getting clusterId {}", reconciliation, e.getMessage());
+            } catch (Exception e) {
```
What kind of exception are we expecting here? I think it would be good to use something more specific exception type if possible.

lga-zurich replied:
As said in the previous comment, the KafkaFuture needs to be handled better anyway; then those exceptions will likely not be a concern anymore.

tombentley replied:
Because of where this is running within `createOrUpdate` I'm guessing it will could fail if the broker aren't ready. Currently that's not really a problem, because the cluster id doesn't change, so most of the time no one would notice. And when creating a cluster the 2nd reconciliation will come around soon enough so the status would only be lacking the id for a short while. If we're going to change the exception handling we'd need to think through the error scenarios in a little more detail. Ignoring the exception in some way is probably the right thing to do, because we don't really want to prevent the rest of the reconciliation from happening. But logging at debug is probably a little too drastic since it's kinda useful for a user to know that we couldn't even describe the cluster. We should go as far as to say the status should be `NotReady` though? @scholzj wdyt?

scholzj replied:
I think there are three separate things. 

First ... the `Exception` seems very generic to me. Catching some more specific types would be better. But I have no clue whether that is possible and what does the Admin API throw. If `Exception` is the best we can do then ok.

Then there are the two issues: How do we log the errors and whether we fail the reconciliation. I agree that the clusterId is not critical enough to fail the reconciliation. But maybe then we should bump the log messages to WARN to make sure they get noticed? Otherwise we might one day find out it never works but nobody notices it. That is at least unless it fails for some reason too often and we do not want to confuse users with the WARN messages.

As for the placement, I didn't noticed it in my review. But it makes sense to have it only after the Kafka pod readiness check.

scholzj commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaStatusTest.java:
```diff
@@ -998,6 +998,42 @@ public void testInitialStatusOnOldResource() throws ParseException {
     }
 
+    @Test
+    public void testKafkaClusrerIdInStatus(VertxTestContext context) throws ParseException {
```
I think this tests only one small aspect of this. I think this test would be much more useful, if you mock the Kafka Admin client and have that return a mocker ClusterID. And then in the reconciliation you actually should let the `kafkaGetClusterId` run as it is in the code. To have it tested as well.

lga-zurich replied:
Agreed, I'll follow your pointer below, the test I include didn't test much at all.

scholzj reviewed (changes requested):
Thanks for the PR. It looks mostly good, but I left few comments. Also some things which seem to be missing:
* Please add a record about this to the CHANGELOG.md file
* After changing the API (custom resource definitions) you have to run `make crd_install` to copy them into all kind of places. That is what probably what failed the tests.
* I'm a bit surprised this did not needed to be mocked for any of the other tests. I wonder if it is because you just let it pass instead of failing the reconciliation. But I think it would be good to have this mocked. I think you just need to set it in the `ResourceUtils` class, where it already mocks the AdminClient and DescribeClusterResult (see here: https://github.com/strimzi/strimzi-kafka-operator/blob/08e4b9862489c423e06905479988ccbdd9974116/cluster-operator/src/test/java/io/strimzi/operator/cluster/ResourceUtils.java#L642)

lga-zurich commented:
> Thanks for the PR. It looks mostly good, but I left few comments. Also some things which seem to be missing:
> 
> * Please add a record about this to the CHANGELOG.md file

Sure, no problem.

> * After changing the API (custom resource definitions) you have to run `make crd_install` to copy them into all kind of places. That is what probably what failed the tests.

ok, it all compiled and run fine on my laptop using Minikube using `make docker_build`. I'll do that and check.

> * I'm a bit surprised this did not needed to be mocked for any of the other tests. I wonder if it is because you just let it pass instead of failing the reconciliation. 

I didn't want to fail this on purpose, as I didn't want to have this blocking any critical path I might have been unaware of. 

>But I think it would be good to have this mocked. I think you just need to set it in the `ResourceUtils` class, where it already mocks the AdminClient and DescribeClusterResult (see here: https://github.com/strimzi/strimzi-kafka-operator/blob/08e4b9862489c423e06905479988ccbdd9974116/cluster-operator/src/test/java/io/strimzi/operator/cluster/ResourceUtils.java#L642
>   )

That's a good pointer, thanks! I'll check that out.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -296,6 +300,7 @@ public KafkaAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability pfa,
                 .compose(state -> state.kafkaRoutes())
                 .compose(state -> state.kafkaIngresses())
                 .compose(state -> state.kafkaInternalServicesReady())
+                .compose(state -> state.kafkaGetClusterId())
```
If we're going to change the exception handling, perhaps this should run last, or at least after `kafkaPodsReady`.

tombentley reviewed (commented):
Thanks for the PR! We need to think through how the error handling should work, but otherwise this is a worthwhile feature.

lga-zurich commented:
Hi @scholzj @tombentley 
I addressed your comments like follows:
- Updated the CHANGELOG.md
- Made obtaining the clusterId non-blocking
- Elevated the potential KafkaException when obtaining the Kafka Admin client to WARN
- Mocked the clusterId as suggested by @scholzj 
- Moved `kafkaGetClusterId()` after services are up and running as suggested by @tombentley

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1954,6 +1958,45 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
+         * @return
+         */
+        Future<ReconciliationState> kafkaGetClusterId() {
+            if (kafkaStatus.getClusterId() != null) {
+                // Cluster Id does not change
+                return Future.succeededFuture(this);
+            } else {
```
The status starts empty every time ... so I would normally not expect this to every happen. Is it really needed?

lga-zurich replied:
ok, I will get rid of it.

scholzj commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaStatusTest.java:
```diff
@@ -1153,4 +1189,18 @@ public MockModelWarningsStatusKafkaAssemblyOperator(Vertx vertx, PlatformFeature
+        }
+
+        @Override
+        Future<Void> reconcile(ReconciliationState reconcileState)  {
+            reconcileState.kafkaStatus.setClusterId("CLUSTERID");
+
+            return Future.succeededFuture();
+        }
```
Since you now mocked the cluster-id in the Admin provider, I do not think ou need this ... you should really just call the actual `kafkaGetClusterId` method IMHO which should set it on its own.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:
```diff
@@ -1954,6 +1958,45 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
+                return adminClientSecrets()
+                    .compose(compositeFuture -> {
+                        log.debug("{}: Attempt to get clusterId", reconciliation);
+                        try {
+                            String bootstrapHostname = KafkaResources.bootstrapServiceName(this.name) + "." + this.namespace + ".svc:" + KafkaCluster.REPLICATION_PORT;
+                            log.debug("{}: Creating AdminClient for clusterId using {}", reconciliation, bootstrapHostname);
+                            Admin kafkaAdmin = adminClientProvider.createAdminClient(bootstrapHostname, compositeFuture.resultAt(0), compositeFuture.resultAt(1), "cluster-operator");
+                            Promise<ReconciliationState> resultPromise = Promise.promise();
+                            vertx.createSharedWorkerExecutor("kubernetes-ops-pool").<ReconciliationState>executeBlocking(
+                                future -> {
+                                    try {
+                                        kafkaStatus.setClusterId(kafkaAdmin.describeCluster().clusterId().get());
+                                        future.complete(this);
+                                    } catch (Throwable e) {
+                                        future.fail(e);
+                                    }
+                                },
+                                true,
+                                resultPromise);
+                            return resultPromise.future();
+                        } catch (KafkaException e) {
+                            log.warn("{}: Kafka exception getting clusterId {}", reconciliation, e.getMessage());
+                        }
+
+                        return Future.succeededFuture(this);
```
I wonder if this does what you want it to do. I think this code would do the warning and return the succeeded future only in case the `adminClientProvider.createAdminClient(bootstrapHostname, compositeFuture.resultAt(0), compositeFuture.resultAt(1), "cluster-operator");` fails. But not when `kafkaAdmin.describeCluster().clusterId().get()` fails. When `kafkaAdmin.describeCluster().clusterId().get()` fails the whole reconciliation will fail.

If we really just want to pass the reocnciliation when this fails as discussed before, maybe you need to turn it around and have the whole try-catch block inside the `executeBlocking` section.

tombentley replied:
`kafkaAdmin` doesn't seem to get closed.

lga-zurich replied:
@scholzj ok, I'll move everything into the `executeBlocking`
@tombentley good catch!

scholzj reviewed (commented):
Looks mostly good. Just some more comments.

scholzj commented:
@lga-zurich Any chance you can have a look at the comments so that we can merge it before releasing 0.21.0?

lga-zurich commented:
> @lga-zurich Any chance you can have a look at the comments so that we can merge it before releasing 0.21.0?

I just came back from holidays, but this shouldn't take too long. When is the deadline?

scholzj commented:
> I just came back from holidays, but this shouldn't take too long. When is the deadline?

I think we would want to start the release end of this week / early next week. There are many users who want Kafka 2.7.0. I do not think from Strimzi perspective it is a big issue if this PR slips into a later release. So I basically wanted to let know about the release plan. But from our side I do not think there is any issue if we take it slowly and it gets only into the next release.

scholzj commented:
Actually I see you already pushed some changes ... it looks like the CI failed on the checkstyle. Probably some forgotten import or something.

scholzj commented on .checkstyle/checkstyle.xml:
```diff
@@ -103,7 +103,7 @@
         <module name="ClassFanOutComplexity">
             <!-- default is 20 -->
-            <property name="max" value="44"/>
+            <property name="max" value="45"/>
```
Can you instead of changing this use the `@SuppressWarnings({"checkstyle:ClassFanOutComplexity"})` annotation in the code which is failing?

lga-zurich replied:
No problem, I felt a bit uncomfortable changing that anyway. I already had to hardcode the names of the secrets to avoid using more imports, which is also wrong.

scholzj commented:
/azp run regression

tombentley reviewed (approved):
LGTM.

scholzj commented:
Thanks for the PR.

lga-zurich commented:
Cheers @scholzj @tombentley
